contrib/immich.lua

```diff
@@ -0,0 +1,335 @@
+  if (err ~= 200) then 
+    if err == 401 then
+      extra_data.error = "Authentication error. Check your Immich API key in LUA settings."
+    elseif res ~= nil and res.message ~= nil then
+      extra_data.error = res.message
+    else 
+      extra_data.error = "Error contacting Immich server: HTTP "..err
+    end
+    return {} 
+  end
+  if (res.existingIds ~= nil) then 
+    for i,id in ipairs(res.existingIds) do
+      extra_data.images_existence[id] = true
+    end
+  end
+
+  extra_data.album_assets = {}
+  extra_data.remote_albums = {}
+  
+  local res_albums, err_albums = call_immich_api("GET","albums")
+ 
+  if err_albums == 200 then
+    for _,album in ipairs(res_albums) do
+      extra_data.remote_albums[album.albumName] = album.id
+    end
+  end
+  
+  return images
+end
+
+local function iso_exif_datetime_taken(image) 
+  local yr,mo,dy,h,m,s = string.match(image.exif_datetime_taken, "(%d-):(%d-):(%d-) (%d-):(%d-):(%d+)")
+  return os.date("!%Y-%m-%dT%H:%M:%S",os.time{year=yr, month=mo, day=dy, hour=h, min=m, sec=s})
+end
+
+local function replace_image(image,filename,device_id,asset_id) 
+  local date = iso_exif_datetime_taken(image)
+  local form_data = {
+    deviceAssetId=tostring(image.id),
+    deviceId=device_id,
+    fileCreatedAt=date,
+    fileModifiedAt=date,
+    assetData={
+      filename=df.get_filename(filename),
+      file=io.open(filename)
+    }
+  }
+  local res,err = call_immich_api("PUT","assets/"..asset_id.."/original",form_data,"multipart/form-data")
+  if err == 200 then
+    return asset_id
+  end
+  return nil
+end
+
+local function upload_image(image,filename,device_id) 
+  local date = iso_exif_datetime_taken(image)
+  local form_data = {
+    deviceAssetId=tostring(image.id),
+    deviceId=device_id,
+    fileCreatedAt=date,
+    fileModifiedAt=date,
+    assetData={
+      filename=df.get_filename(filename),
+      file=io.open(filename)
+    }
+  }
+  local res,err = call_immich_api("POST","assets",form_data,"multipart/form-data")
+  if err == 201 then
+    return res.id
+  end
+  return nil
+end
+
+local function store_image(storage,image,format,filename,number,total,high_quality,extra_data)
+  local asset_id,replaced = false
+  if (extra_data.images_existence[tostring(image.id)]) then
+    local res_search,err_search = call_immich_api("POST","search/metadata",{deviceId=extra_data.device_id,deviceAssetId=tostring(image.id),size=1})
+    if (err_search == 200) then 
+      if (res_search.assets.count >= 1) then
+        replaced = true
+        asset_id = replace_image(image,filename,extra_data.device_id,res_search.assets.items[1].id)
+      else 
+        asset_id = upload_image(image,filename,extra_data.device_id)
+      end
+    end
+  else
+    asset_id = upload_image(image,filename,extra_data.device_id)
+  end
+
+  if asset_id == nil then 
+    extra_data.error = "Error uploading some image"
+    return
+  end
+
+  if not replaced then
+    local album_name = title_widget.text
```

Comment:
This line generates an error because the title_widget doesn't get defined until later.  The above comment about inserting the `local title_widget` line fixes that.